### PR TITLE
Set workspace-jupyterlab-python to use 1001:1001, launch via gosu

### PR DESCRIPTION
**Pull request by echuber2**
Fixes https://github.com/PrairieLearn/PrairieLearn/issues/9828

- In the install script, we now change the default jovyan 1000:100 to 1001:1001. This also involved changing the default group "users" gid 100 to 1001 but not renaming the group.
- Use gosu to step down to 1001:1001 when launched as 0:0 (in local testing, for example).
- Tried using uv to accelerate pip install, since the build is pretty big/slow with so much happening all in the install script. (Maybe we should break out more Docker layers for caching, but the current synchronized apt upgrade with each change to pip packages may be desired.)

If anybody objects to the `uv` usage, it's no problem to back that out. The gosu stuff doesn't depend on it.

### CODE REVIEW

**Comment by echuber2** [comment]
Side note about a failed alternative: I did experiment with simply adding an additional user and group 1001:1001 to the image, and also adding this user to the `users` group that Jupyter has by default, since this is normally something they would recommend for this use case. However, this only seems to help if the container is run as 1001:100 for some reason, and running as 1001:1001 after doing that breaks it worse than before. So, the currently-proposed changes are the simplest thing I can suggest.

**Comment by echuber2** [comment]
As a stopgap until this is merged, I made a post explaining a temporary workaround that at least helps with locally testing the workspace (enabling saving and submitting): https://github.com/PrairieLearn/PrairieLearn/discussions/10250

File: workspaces/jupyterlab-python/install.sh
Comment (nwalters512) [inline workspaces/jupyterlab-python/install.sh:35]: If someone wants to build a customized image on top of this one, would this mean they'd have to use `uv` too, or would using `pip` play nicely here (install into the same global namespace without conflicts and such)?

From my experience with conda/mamba in other images, I'd expect things to work like this, but just wanted to check since we haven't used `uv` before.

**Review by nwalters512 — COMMENTED** [review]
I built this image locally as `prairielearn/workspace-jupyterlab-python:nathan` and updated the `demo/workspace/jupyterlab` question to use it.

I tried testing this locally in a few ways (all on macOS):

- Run `make dev` + `make dev-wokspace-host` as my normal user with no special config. When I try to launch a workspace, that fails with `rror starting workspace 4 EPERM: operation not permitted, chown '/Users/nathan/.prairielearn/jobs/workspace-4-1/current-3dc316e9-4714-46a6-9c4d-b573af7f33d6'` ([source code here](https://github.com/PrairieLearn/PrairieLearn/blob/6a77a92af4a067efc400bf0a328cf10b964a7752/apps/prairielearn/src/lib/workspace.ts#L396))
- I tried the same, but running as root (`sudo make dev` + `sudo make dev-workspace-host`). Attempting to launch a workspace failed with `Max startup time exceeded`.
- I ran it again, but this time with `"workspaceJobsDirectoryOwnerUid": 1001, "workspaceJobsDirectoryOwnerGid": 1001` in my config.
  - As my normal user, this fails the same as above (`operation not permitted, chown '/Users/nathan/.prairielearn/jobs/workspace-7-1/current-02573117-f7a1-4e30-91fe-69d9e6c9d836'`)
  - As root (with `sudo`), this allowed me to launch the workspace, but trying to save a notebook failed with `File Save Error for Workbook.ipynb Permission denied: Workbook.ipynb`.

Maybe I misunderstood what this PR is meant to do. Is this meant to make this workspace work with native local development? Or does this just improve things for folks who are doing local dev with the Docker container?

It'd be helpful to have some testing instructions for this, e.g. "do X with the old image, observe Y, do the same thing with the updated image, observe Z".

**Comment by echuber2** [comment]
> Maybe I misunderstood what this PR is meant to do. Is this meant to make this workspace work with native local development? Or does this just improve things for folks who are doing local dev with the Docker container?

The existing workspace did have the side-referenced issues about working locally but the main reason for the PR was that it has issues on the deployed server too. If I open the example course on the Illinois server and try question `demo/workspace/jupyterlab`, opening the terminal in Jupyter Lab, I experience this:

```
groups: cannot find name for group ID 1001
I have no name!@534c01fb9ee2:~$ id
uid=1001 gid=1001 groups=1001
I have no name!@534c01fb9ee2:~$ pwd
/home/jovyan
I have no name!@534c01fb9ee2:~$ ls -hal
total 12K
drwxr-xr-x. 2 1001 1001 6.0K Sep  3 22:12 .
drwxr-xr-x. 1 root root 4.0K Oct 20  2023 ..
-rw-r--r--. 1 1001 1001 1.9K Sep  3 22:12 Workbook.ipynb
I have no name!@534c01fb9ee2:~$ groups jovyan
jovyan : users
I have no name!@534c01fb9ee2:~$ id jovyan
uid=1000(jovyan) gid=100(users) groups=100(users)
```

So as things are currently (sans PR), jovyan is 1000, and we are running as a no-name 1001. Coincidentally the example question does have `"home": "/home/jovyan"` so the 1001 user ends up owning this directory and the workbook file anyway, but it's a potential pitfall and other libraries or Jupyter features might be breaking with the unmatched user identity.

Once we start stepping down in the entrypoint to run as 1001 locally, we run into the issue locally where the local files might be owned by root, and that's why the chowning would be needed (and should be working) locally. I'll have to try this out on macOS myself to see if I can understand why that's behaving as it is for you.

**Comment by nwalters512** [comment]
Ahh, ok. I thought this was fixing some user-facing error (either locally or in prod), but really this is just about being internally consistent with what user we're running on, which has the nice benefit of fixing the "I have no name!" terminal prompt, is that right?

**Comment by echuber2** [comment]
> Ahh, ok. I thought this was fixing some user-facing error (either locally or in prod), but really this is just about being internally consistent with what user we're running on, which has the nice benefit of fixing the "I have no name!" terminal prompt, is that right?

Besides the "I have no name" thing, there are some other weird side effects I noticed trying to adjust settings in Jupyter with the existing no-name setup, although I'd have to experiment again and take notes. Your issues in macOS sound like probably expected behavior if you're running PL itself as your normal user (since chowning stuff would then fail with privilege escalation), and in the case when you did run PL as root, I'm not sure what happened:

> As root (with sudo), this allowed me to launch the workspace, but trying to save a notebook failed with File Save Error for Workbook.ipynb Permission denied: Workbook.ipynb.

So that's exactly the kind of thing I was hoping to avoid by having 1001:1001 own the home directory and _be_ the jovyan account, and also running the workspace as 1001:1001 whether locally (using gosu) or in prod (by default). I wonder if your local setup is reusing a job directory and running into stale files owned by root, but it seems we had resolved that in the past with a PR that moved stale workspace files out of the way, so I'm not sure. Things seemed to be working for me on Linux using the PL Docker (running as root) plus running the workspace either using 0:0 + gosu or as 1001:1001 directly.

Zac Warham @ZacWarham has started a draft PR for a workspace with Postgres built in, so it may help to be able to base that work on this, or else I might need to patch in similar temporary fixes experimenting with the Postgres setup. That probably won't hurt but if it's eventually redundant, we'll need to remember to circle back and simplify that other work.

**Comment by joelostblom** [comment]
As someone who has recently started using PL I have come across two workflow issues when developing questions locally:

1. My local user doesn't own files created via the UI running from the docker container. Workaround in https://github.com/PrairieLearn/PrairieLearn/discussions/10661
2. Running workspaces via the local docker container involves workarounds and is still suggested to potentially have different effects than when workspaces are run on the US server. 

My understanding is that this PR is aiming to solve (2) above, which I think is great! I wonder if the solution can be made compatible with the workaround for (1) above as well. That way PL would have a local docker workflow that can be used identically to how things would work on the server, which would be very helpful when creating questions locally. 

@echuber2 I saw that you mentioned the following in 
https://github.com/PrairieLearn/PrairieLearn/issues/6998#issuecomment-1407137481 (em mine) 

> There is a related issue (like what #3139 refers to) that **many base images assume the user is 1000, not 1001**. So really, we should be customizing all workspaces to re-ID the default account to 1001 in the image itself (not at launch time), as in the example here: #3170 **Or, we could transition to using 1000:1000 on the server instead of 1001:1001, but now that would break these other adaptations we've done for the 1001 case**. It would help if workspaceOptions in a question info.json could configure which user to become, with the default being 1001 as before. (I think that's better than relying on the last USER setting in the Docker itself, but maybe I'm wrong.) **This would let workspace developers just say to run as 1000 instead of 1001** and skip the modification shown in #3170, while also being backwards-compatible with existing workspaces.

Using user id 1000 instead of 1001 to run workspaces by default sounds appealing for me as this would also work nicely with starting the docker containers as a non root user (which is usually 1000). Do you think this is still a viable solution? Or is there anything inherently preferable with using user 1001 on the server?

**Comment by echuber2** [comment]
With this PR, we're baking in the 1001:1001 uid and gid for jovyan, so baking in 1000:1000 instead would require separate builds for local testing. On the other hand dynamic switching could be done at runtime but this causes a slower startup. Actually, the Jupyter upstream images can do this already, if you run them as root and specify some environment variables. So, for the case of Jupyter in particular, there would be a way to accommodate your request pretty easily and run as 1000 instead.

For other PL workspace images, we would have to add that dynamic behavior ourselves. However there may be cases where this isn't trivial. For example, recent Ubuntu base images already have an "ubuntu" user 1000, and the particular image may add another default user on top of that already (like 1001, coincidentally what we want! This happens with code-server currently, but we don't want to _rely_ on that assumption).

Another thing you might be able to do is create a local group 1001 and add your personal user to that group also. (Or this could be done by creating a user/group 1001:1001 combo locally and, likewise, adding your user 1000 to the 1001 group.) However I don't know if these workarounds will work for all users, depending on their situation. What if a user 1001 already exists on the machine?

Another factor is that in the future, it may be necessary for us to recommend people run Docker/Podman _rootlessly_ in local testing, using user ID namespace remapping. This is what would allow advanced workspaces to give the student fake root, for example. But this also remaps all uid and gid into an upper range virtually, so once again, your default local user 1000 outside the container won't automatically have ownership of those generated files. There would be a way to configure this so that your user does own the remapped range, though, I think.

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh:35]: I thought I had submitted these comments already but it looks like they were in draft only. If there are any doubts about `uv` here then by all means, we can just use pip instead.

**Comment by echuber2** [comment]
> For other PL workspace images, we would have to add that dynamic behavior ourselves. However there may be cases where this isn't trivial.

I may go ahead and try adding this feature to the entrypoint script since it seems like a good time to do it, with the PR still being reviewed.

**Comment by joelostblom** [comment]
Thanks for the explanation and sorry for the late reply. I don't fully understand if the limitations of running the workspaces as user 1000 is because of something inherent to that user ID or because PL chose to use 1001 initially and it would take a lot of work to change everything to user 1000 (which I would then not advocate for).

If this PR makes workspaces function the same locally as on the server, then I think that is great even if it wouldn't allow me to create files via the UI that are owned by my user. I don't create files this way frequently anyways and would be OK running a different command when I need to.

> So, for the case of Jupyter in particular, there would be a way to accommodate your request pretty easily and run as 1000 instead.

Do you already know what this would look like?

> However I don't know if these workarounds will work for all users, depending on their situation. What if a user 1001 already exists on the machine?

Yes, I agree that many of these assumptions seem brittle.

File: workspaces/jupyterlab-python/install.sh
Comment (joelostblom) [reply workspaces/jupyterlab-python/install.sh:35]: > If someone wants to build a customized image on top of this one, would this mean they'd have to use uv too

If we create a base jupyterlab image without any python packages, then that could be the main target for building custom images and make this point less of a restriction. I believe that would also allow those who want to use the built-in mamba installer to  install `gcc` and `graphviz`, which I assume are needed for some of the python packages and not for any core PL/docker functionality?

**Comment by echuber2** [comment]
> Do you already know what this would look like?

@joelostblom For example, I began prototyping something where you could put this in your `info.json`:

```json
  "workspaceOptions": {
    "image": "prairielearn/workspace-jupyterlab-python",
    "port": 8080,
    "home": "/home/jovyan",
    "rewriteUrl": false,
    "gradedFiles": ["Workbook.ipynb"],
    "environment": {
      "LOCAL_UID": "1000"
    }
  },
```

and then when the workspace is run as root (like by default), it would respect `LOCAL_UID` to do what you want. However, then you might commit this edit to the `info.json` to your course and annoy other staff. (In production it would have no effect.)

Another alternative would be to skip this runtime feature and instead, provide explicit instructions about how to customize a local build of the workspace image to bake in the defaults that you prefer.

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh]: I went ahead and removed `uv` so that this won't be a blocker any longer.

Comment (nwalters512) [reply workspaces/jupyterlab-python/install.sh]: I do think the speed improvements of `uv` are intriguing all around (including perhaps for the PL base image itself), but after thinking about this I'd rather limit the number of things that this PR is changing. I see you reverted to `pip` in a recent comment, I think that's the right move for now.

File: workspaces/jupyterlab-python/install.sh
Comment (nwalters512) [inline workspaces/jupyterlab-python/install.sh:7]: Nit: is this necessary? I don't see this in other workspaces and since this is always called from the Dockerfile with `USER root` preceding it, I don't see how this could end up being run as non-root.

File: workspaces/jupyterlab-python/install.sh
Comment (nwalters512) [inline workspaces/jupyterlab-python/install.sh:27]: Nit: what does it mean to "test gosu"?

File: workspaces/jupyterlab-python/install.sh
Comment (nwalters512) [inline workspaces/jupyterlab-python/install.sh:34]: Nit: why the change from `pip3` to `pip`?

**Comment by echuber2** [comment]
For now I removed the experimental uv install step (at the cost of a slower build) and perhaps the refactoring to base image or adding the custom local user features could be separate PRs. I just wanted to bump this because some folks mention permissions issues with Jupyter once or twice a week on the slack but it seems like this could have been a solved problem a year or two ago. :sweat_smile:

**Review by nwalters512 — APPROVED** [review]
A few nits, but none of these are blockers. Approved; thank you for your patience and persistence with this!

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh:7]: It's just an assertion to avoid configuration errors. It could be removed (most of the following script would fail if the condition isn't met anyway).

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh:27]: The gosu installation instructions [demonstrated doing this](https://github.com/tianon/gosu/blob/master/INSTALL.md) to check that it's working as intended (although it doesn't seem comprehensive).

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh:34]: With Python 2 in the past, "pip3" seems like an anachronism. It may not matter as long as the distro has both. On Ubuntu it looks like these are installed as two identical files (not necessarily hardlinked).

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh]: @joelostblom It looks like just installing python3 and python3-pip on Ubuntu pulls in gcc and g++ (among other things). I think the other PL images ran into some trouble getting graphviz installed depending on the distro (for the latest VS Code image, mamba was used to install the binary).

File: workspaces/jupyterlab-python/install.sh
Comment (nwalters512) [reply workspaces/jupyterlab-python/install.sh:27]: This feels a little idiosyncratic; none of our other Dockerfiles test things that they install, AFAIK. It's your call whether you think this is valuable enough to leave here.

File: workspaces/jupyterlab-python/install.sh
Comment (nwalters512) [reply workspaces/jupyterlab-python/install.sh:7]: I'll leave that call to you; approval stands.

File: workspaces/jupyterlab-python/install.sh
Comment (nwalters512) [reply workspaces/jupyterlab-python/install.sh:34]: In our other Dockerfiles and setup scripts, we either use `pip3` or `python3 -m pip ...`. I'd like to remain as consistent as possible (though I acknowledge that we're already inconsistent by having two ways of doing this).

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh]: All right, I went ahead and removed it.

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh]: I just removed it. It would have made more sense to test whether `jovyan` had the expected uid:gid but it's an arbitrary thing to unit test out of all the other things in the install process.

File: workspaces/jupyterlab-python/install.sh
Comment (echuber2) [reply workspaces/jupyterlab-python/install.sh]: Went ahead and changed it back to pip3.

**Pull request merged**
